### PR TITLE
Merge many simple parts of GAP-HPC library

### PR DESCRIPTION
**Pull request by ChrisJefferson**
This merges many simple parts of the GAP-HPC library back into GAP.

This commit builds on #231 , after that is merged all changes here are (hopefully) very simple.

### CODE REVIEW

**Comment by alex-konovalov** [comment]
@ChrisJefferson if you will look at `ratfunul.gi`, it will be modified by PR #243 which may go first.

File: lib/integer.gi [hunk 1]
```diff
@@ -980,25 +992,37 @@ BindGlobal( "IsProbablyPrimeIntWithFail", function( n )
...
     if n < 0         then n := -n;       fi;
     if n in Primes   then return true;   fi;
+    atomic readonly Primes2 do
     if n in Primes2  then return true;   fi;
-    if n in ProbablePrimes2  then return fail;   fi;
+od;
+atomic readonly ProbablePrimes2 do
+if n in ProbablePrimes2  then return fail;   fi;
+od;
     if n <= 1000     then return false;  fi;
```
Comment (fingolfin) [inline lib/integer.gi]: That's bad code formatting -- but I assume this is the same as in the HPC-GAP lib? Ah well. I also copied some badly formatted C code in PR #239 ;)

File: lib/matint.gi [hunk 2]
```diff
@@ -237,7 +237,7 @@ local k,g,b,ii,m1,m2,t,tt,si,n,m,i,j,r,jj,piv,d,gt,tmp,A,T,TT,kk;
 end);
 
 BindGlobal("BITLISTS_NFIM",
-  [ [ false, false, false, false, false ], [ true, false, false, false, false ],
+MakeImmutable( [ [ false, false, false, false, false ], [ true, false, false, false, false ],
     [ false, true, false, false, false ], [ true, true, false, false, false ],
```
Comment (fingolfin) [inline lib/matint.gi]: One really notices how some code was clearly converted before the backtick notation was introduced... I guess we should ignore that for now, but once the libs are merged, I guess it would be good to go over that and use backticks in more places.

Comment (ChrisJefferson) [inline lib/matint.gi]: I agree. Now you've mentioned this, I'm wondering if I should go and backtick HPC-GAP, and then re-make this commit, or just do the merge and then backtick later (that's my temptation to be honest, mainly to avoid re-doing this merge)

**Comment by alex-konovalov** [comment]
Yes, some code was indeed converted before backtick was introduced. Backtick is only a shortcut, so I don't see any harm in merging first and then backticking later.

**Comment by alex-konovalov** [comment]
#231 now merged. Perhaps we could wait for Thursday round of nightly tests with Jenkins, and then merge this on Friday too.

**Comment by fingolfin** [comment]
@alex-konovalov I cannot access the GAP Jenkins (its using an invalid SSL certificate, and even if I bypass that grudingly, it doesn't like the username / password I have stored), so I am not sure what the results are... But hopefully good? 
I'd like to see this merged this ASAP. :)

**Comment by alex-konovalov** [comment]
@fingolfin : well, we have a problem. `make teststandard` fails, and I haven't yet figured out why: the log just says
```
rev: stdin: make: *** [teststandard] Error 1
```
but both log files contain the "total" line, showing that all tests were completed.

**Comment by ChrisJefferson** [comment]
I can't get teststandard to run -- too many packages fail to load.

However, in the new testing framework, if any tests fail, 'make teststandard' should return fail. Are some tests failing?

**Comment by alex-konovalov** [comment]
@ChrisJefferson,

Hmm... one should be able to run it even if for some packages `LoadPackage(...)` returns fail. What exactly are you doing? Also, `make teststandard` is not a function so it can't return fail - it just produces two log files in `dev/log`, so I am confused by the 2nd paragraph too...

**Comment by alex-konovalov** [comment]
@ChrisJefferson: P.S: and yes - nothing is visibly failing. The "total" line is shown in the log file, meaning that the test was performed fully (no segfault), and also there are no diffs in tests, so none of them are failing.

**Comment by alex-konovalov** [comment]
Looked a bit more at the build history. First time this happened on Wednesday night, with the last change in the master branch snapshot being merged PR #239  7f5d9f4e1161da2043fde0f4c92a787b3be88fef. Also, two new packages picked up earlier that day in the morning were homalg 2015.09.21 and Semigroups 2.6. Now, after they were picked up, there was a quick test performed in just one configuration based on the commit  3dad1f719b96684b0b1bb5166b644fee14a8425a (merged PR #241) and the problem was already there. This means that PR #231 which we are evaluating, was not yet there at this point and can not be blamed. Therefore, answering @fingolfin's question, I think it is fine to merge #249 i.e. this PR then.
 
Still an open problem why `make teststandard` fails however. My next step is to test it with the collection of packages that worked a day before the failure. However, the newly picked up packages were tested also with stable-4.7 branch, and worked fine there, so if the problem is in packages, then it should be some incompatibility with recent changes in the master. Otherwise, it could be something in PR #241 - before that, there was PR #242 merged, and it worked fine...

**Comment by ChrisJefferson** [comment]
When I run 'teststandard', somewhere during the package loading GAP enters the break loop. I may have an out-of-date set of packages, or a different set of packages compiled.

**Comment by alex-konovalov** [comment]
I have tried on Ubuntu with pkg directory symlinked from 4.7.8 and can see only warnings from packages coming from new variadic functions, but the test actually runs. Please show which errors do you see, maybe they look familiar.

**Comment by alex-konovalov** [comment]
> However, in the new testing framework, if any tests fail, 'make teststandard' should return fail. Are some tests failing?

@ChrisJefferson: I now understood what you mean - this is the exit code, which has been fixed in PR #242. When it had been merged, the test worked fine, but next time the Semigroups package was picked up and it breaks the test by two diffs reported in #258. Now, the script used by Jenkins calls `make teststandard`, it correctly fails and Jenkins just exits immediately. So - good news - #231 is cleared, and we may now merge this PR too. I will modify Jenkins scripts to not to exit when "make test..." fails since I want to do some post-processing there.

**Comment by alex-konovalov** [comment]
FYI, I was trying to measure how large is still the difference between two branches:
```
$ git diff --shortstat master:src hpcgap-default:src 
 91 files changed, 20739 insertions(+), 2260 deletions(-)
$ git diff --shortstat master:lib hpcgap-default:lib
 118 files changed, 30425 insertions(+), 22831 deletions(-)
```

**Comment by alex-konovalov** [comment]
@ChrisJefferson: there is just one problem showing up in Jenkins tests in the check of manual examples:
```
============================================================
########> Diff in [ "./../../lib/ctbl.gd", 3458, 3509 ]
# Input is:
CharacterTableDisplayDefaults.User:= rec(
      powermap:= "ATLAS", centralizers:= "ATLAS", chars:= false );;
# Expected output:
# But found:
Error, Records Assignment: <rec> must be a mutable record
########
########> Diff in [ "./../../lib/ctbl.gd", 3458, 3509 ]
# Input is:
Display( CharacterTable( "A5" ) );
# Expected output:
A5

   60  4  3  5  5

p      A  A  A  A
p'     A  A  A  A
   1A 2A 3A 5A B*

# But found:
A5

    2  2  2  .  .  .
    3  1  .  1  .  .
    5  1  .  .  1  1

      1a 2a 3a 5a 5b
   2P 1a 1a 3a 5b 5a
   3P 1a 2a 1a 5b 5a
   5P 1a 2a 3a 1a 1a

X.1     1  1  1  1  1
X.2     3 -1  .  A *A
X.3     3 -1  . *A  A
X.4     4  .  1 -1 -1
X.5     5  1 -1  .  .

A = -E(5)-E(5)^4
 = (1-Sqrt(5))/2 = -b5
########
########> Diff in [ "./../../lib/ctbl.gd", 3458, 3509 ]
# Input is:
Unbind( CharacterTableDisplayDefaults.User );
# Expected output:
# But found:
Error, Record Unbind: <rec> must be a mutable record
########
=========OUTPUT END: testmanuals with autoloaded packages=========
```
